tests/fuzzers/nodestatemachine/nodestatemachine-fuzzer.go

```diff
@@ -0,0 +1,479 @@
+		if fn := ops[op]; fn != nil {
+			fn(n, true)
+		}
+	}
+
+	var allFlags nodestate.Flags
+	for _, f := range flags {
+		allFlags = allFlags.Or(f)
+	}
+	var offState state
+
+	var flagSubs [16]struct {
+		a, b, c int
+		ops     [8]byte
+	}
+	shift := byte(42)
+	for i := range flagSubs {
+		x := f.randomInt(flagCount * flagCount * flagCount)
+		flagSubs[i].a = x % flagCount
+		x /= flagCount
+		flagSubs[i].b = x % flagCount
+		x /= flagCount
+		flagSubs[i].c = x % flagCount
+		var subops [8]byte
+		for j := range subops {
+			flagSubs[i].ops[j] = f.randomByte() + shift
+			shift += 77
+		}
+	}
+
+	var fieldSubs [fieldCount]struct {
+		ops [8]byte
+	}
+	for ii := range fields {
+		var subops [8]byte
+		for j := range subops {
+			fieldSubs[ii].ops[j] = f.randomByte() + shift
+			shift += 77
+		}
+	}
+
+	addSubs := func() {
+		ns.SubscribeState(allFlags.Or(s.OfflineFlag()), func(n *enode.Node, oldState, newState nodestate.Flags) {
+			o1, o2 := oldState.HasAll(s.OfflineFlag()), newState.HasAll(s.OfflineFlag())
+			if o1 != o2 {
+				var st nodestate.Flags
+				if o1 {
+					st = newState
+				} else {
+					st = oldState
+				}
+				var m uint64
+				for _, f := range flags {
+					m += m
+					if st.HasAll(f) {
+						m++
+					}
+				}
+				offState[int(n.ID()[0])].mask = m
+			}
+		})
+
+		for _, sub := range flagSubs {
+			ns.SubscribeState(nodestate.MergeFlags(s.OfflineFlag(), flags[sub.a], flags[sub.b], flags[sub.c]), func(n *enode.Node, oldState, newState nodestate.Flags) {
+				if oldState.Or(newState).HasAll(s.OfflineFlag()) {
+					return
+				}
+				callbackCount++
+				if callbackCount > cbLimit {
+					return
+				}
+				j := 0
+				if newState.HasAll(flags[sub.a]) {
+					j += 1
+				}
+				if newState.HasAll(flags[sub.b]) {
+					j += 2
+				}
+				if newState.HasAll(flags[sub.c]) {
+					j += 4
+				}
+				runsub(n, sub.ops[j])
+			})
+		}
+
+		for ii, fi := range fields {
+			fieldIdx := ii
+			ns.SubscribeField(fi, func(n *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+				if state.HasAll(s.OfflineFlag()) {
+					v := oldValue
+					if v == nil {
+						v = newValue
+					}
+					var vv uint16
+					if uv, ok := v.(uint64); ok {
+						vv = 256 + uint16(byte(uv))
+					}
+					if uv, ok := v.(uint16); ok {
+						vv = 512 + uint16(byte(uv))
+					}
+					offState[int(n.ID()[0])].fields[fieldIdx] = vv
+					return
+				}
+
+				callbackCount++
+				if callbackCount > cbLimit {
+					return
+				}
+				var j int
+				if newValue != nil {
+					if u, ok := newValue.(uint64); ok {
+						j = int(u%7) + 1
+					} else {
+						j = int(u%7) + 1
+					}
+				}
+				runsub(n, fieldSubs[fieldIdx].ops[j])
+			})
+		}
+	}
+	for i := range ops {
+		b := f.randomByte()
+		if b+byte(i*137) < 4 {
+			break
+		}
+		b2 := f.randomByte()
+		u := uint(b) + uint(b2)<<8
+		ops[i] = func(n *enode.Node, sub bool) {
+			u := u
+			optype := u % 5
+			u /= 5
+			nn := n
+			shift := u % 4
+			u /= 4
```

Comment:
What is this? What's happening?

Reply:
I tried to use the quasi-random input string economically :) This part takes two bytes of input and transtlates it to a single operation. These are then assigned to byte values. Subscription callbacks, `ns.Operation` batches and the "top level" activity are composed as sequences of these single operations. It is similar to what `randomInt` does, I just do the modulus thing multiple times. Actually I could have used multiple `randomInt` calls (one for each choice) but I thought this is more efficient since the total amount of information needed to construct a single operation is less than two bytes (so I'm not re-using information, just not wasting it either).